storey/writers.py

```diff
@@ -15,58 +15,81 @@
+                    new_value = event.body[column]
+                new_data.append(new_value)
+
+    @staticmethod
+    def _get_column_data_from_list(new_data, event, original_data, columns, metadata_columns):
+        data_cursor = 0
+        for column in columns:
+            if column in metadata_columns:
+                metadata_attr = metadata_columns[column]
+                new_value = getattr(event, metadata_attr)
+                new_data.append(new_value)
+            else:
+                new_data.append(original_data[data_cursor])
+                data_cursor += 1
+        return data_cursor
 
     def _event_to_writer_entry(self, event):
         data = event.body
         if isinstance(data, dict):
-            if self._first_event and self._infer_columns_from_data:
+            if self._infer_columns_from_data:
                 self._columns.extend(data.keys())
                 self._columns.sort()
-
-            if self._columns:
-                new_data = []
-                for column in self._columns:
-                    if column in self._metadata_columns:
-                        metadata_attr = self._metadata_columns[column]
-                        new_value = getattr(event, metadata_attr)
-                    elif column in self._rename_columns:
-                        new_value = data[self._rename_columns[column]]
-                    else:
-                        new_value = data[column]
-                    new_data.append(new_value)
-                data = new_data
-
-        elif isinstance(data, list) and self._columns and self._metadata_columns:
-            new_data = []
-            data_cursor = 0
-            for column in self._columns:
-                if column in self._metadata_columns:
-                    metadata_attr = self._metadata_columns[column]
-                    new_value = getattr(event, metadata_attr)
-                    new_data.append(new_value)
-                else:
-                    new_data.append(data[data_cursor])
-                    data_cursor += 1
-            data = new_data
+                self._infer_columns_from_data = False
+            data = []
+            self._get_column_data_from_dict(data, event, self._columns, self._metadata_columns, self._rename_columns)
+            self._get_column_data_from_dict(data, event, self._index_cols, self._metadata_index_columns, self._rename_index_columns)
+        elif isinstance(data, list):
+            sub_metadata = bool(self._columns) and bool(self._metadata_columns)
+            sub_index_metadata = bool(self._index_cols) and bool(self._metadata_index_columns)
+            if sub_metadata or sub_index_metadata:
+                data = []
+                cursor = self._get_column_data_from_list(data, event, event.body, self._index_cols, self._metadata_index_columns)
+                self._get_column_data_from_list(data, event, event.body[cursor:], self._columns, self._metadata_columns)
         return data
```

Comment:
if it's neither you just return the body and ignore the columns etc the user requested?
Better raise an error saying you can only use the `columns, infer_columns_from_data, index_cols` if your event is either dict or list